packages/@aws-cdk/aws-iotevents/lib/expression.ts

```diff
@@ -70,6 +112,6 @@ class BinaryOperationExpression extends Expression {
   }
 
   public evaluate() {
-    return `${this.left.evaluate()} ${this.operator} ${this.right.evaluate()}`;
+    return `(${this.left.evaluate()} ${this.operator} ${this.right.evaluate()})`;
```

Comment:
Why this change (to add the parenthesis)?

Reply:
Right, but this is not the way to solve this I think 🙂.

I think we should add a `priority` property to all of the expression classes, and each expression will have a different one (`and` will have `2`, `or` will have `1`, etc.). Then, if a sub-expression has a lower-priority then its parent, we will add the parenthesis, if not - we won't.

Hope this makes sense!

Reply:
I agree that the solution (adding priority) can solve the problem.
But I think the first solution I wrote would certainly evaluate them in the order intended by the developer with less code. Because developers don't write `Expression.and(false, Expression.or(true, true))` for they intend to say `(false && true) || true`.

The second solution provide adding less parenthesis automatically like linter auto fixing. But It may be seem as magic. When expressions not work, developers may doubt parenthesis (whichiever the parenthesis caused or not).
The expressions first solution genarated will more descriptive for order of evaluation I think (but verbose?).

Above is my thinking.
Whichever solution we choose, I will spare no effort!

I wish you a good vacation! 😄

Reply:
@yamatatsu  Hmm. But don't the extra parenthesis actually inhibit readability?

In your example, `Expression.and(false, Expression.or(true, true))` will be rendered as `(false && (true || true))`. Obviously, we can make it `false && (true || true)` without changing the meaning. And it would be much more pronounced if the expression was more complex (every binary operation would add their own parenthesis).

So I feel like it's the extra parenthesis that would cause the developers to not like these generated expressions 🙂.

And a side benefit? It will not cause any changes to the existing templates, so will be very nice for customers when upgrading the CDK version once this lands 🙂.

Can you at least give it a shot with the `priority` attribute, and see how it feels? If it's very complicated, I'll let this go, and live with the extra parenthesis, but I'd love to at least give it a shot 🙂.

Reply:
OK! I'll implement it👍🏻

Reply:
@skinny85 
I've done!